packages/shared/lib/core/wallet/actions/activities/linkTransactionsWithClaimingTransactions.ts

```diff
@@ -0,0 +1,67 @@
+import { IAccountState } from '@core/account'
+import { IProcessedTransaction } from '@core/wallet/interfaces'
+import { isOutputAsync } from '@core/wallet/utils/outputs/isOutputAsync'
+import { get } from 'svelte/store'
+import { addClaimedActivity, claimedActivities } from '../../stores'
+
+export function linkTransactionsWithClaimingTransactions(
+    transactions: IProcessedTransaction[],
+    account: IAccountState
+): IProcessedTransaction[] {
+    const resultingTransactions = []
+    const transactionsIncludedAsClaimingTransactions = []
+
+    const claimedAccountActivities = get(claimedActivities)?.[account.id]
+
+    for (const transaction of transactions.sort((t1, t2) => (t1.time > t2.time ? 1 : -1))) {
+        if (transactionsIncludedAsClaimingTransactions.includes(transaction.transactionId)) {
+            continue
+        } else if (transaction.outputs.some((_output) => !isOutputAsync(_output))) {
+            resultingTransactions.push(transaction)
```

Comment:
Currently we are looping through all transactions at the top level, and then if we find an async transaction then we are looping through all transactions again.

I think it would be more efficient to do the linking in this conditional. So you loop through the maximum list only once, and then only loop through transactions added to the async transactions.

Three conditionals being:
1. Do we already have cached data if so then link and update lists straight away
2. Is the transaction async? If so add it to the async transactions list but it is still waiting to be linked.
3. Id transaction is not async then we can check the list of async transactions to see if it is matches this one.

We could also optomise if we made the assumption that an async transaction is only going to a many to 1 link with transactions because we can remove it from the async list after its linked so we don't check it again.